### PR TITLE
Add quote validation

### PR DESCRIPTION
**Pull request by berroar**
### PR Context
Added validation to check for the use of dumb quotes within lines of text. We should only be using smart quotes within our schema text. See http://smartquotesforsmartpeople.com/

### How to Review
Check the new tests created that find dumb quotes.
Run the validation tests against the new test schema.

### CODE REVIEW

**Comment by berroar** [comment]
[Inconsistent quotes used in Welsh translations](https://trello.com/c/6LEeu41c/3228-inconsistent-quotes-used-in-welsh-translations)

File: app/validation/validator.py [hunk 1]
```diff
@@ -1361,6 +1363,93 @@ def _validate_placeholder_transforms(self, transforms, block_id):
...
+                    'Found dumb quotes in question/answer guidance for id {}'.format(sub_block_id)))
+            if ('description' in sub_block_contents[0]) and \
+                    (self._search_dumb_quotes(sub_block_contents[0]['description'])):
+                errors.append(self._error_message(
+                    'Found dumb quotes in question/answer guidance for id {}'.format(sub_block_id)))
+            if 'list' in sub_block_contents[0]:
+                for item in sub_block_contents[0]['list']:
+                    if self._search_dumb_quotes(str(item)):
+                        errors.append(self._error_message(
+                            'Found dumb quotes in question/answer guidance for id {}'.format(sub_block_id)))
```
Comment (ajmaddaford) [inline app/validation/validator.py]: In `eq-translations` we search for all text elements in a schema (have a look at the `no_context_pointers` and `context_pointers` methods). We should look to extract that into something common that we can use here.

File: app/validation/validator.py [hunk 2]
```diff
@@ -1361,6 +1363,93 @@ def _validate_placeholder_transforms(self, transforms, block_id):
...
+            question_title = question['title']['text']
+
+        if self._search_dumb_quotes(question_title) is not None:
+            errors.append(self._error_message(
+                'Found dumb quotes in question title for question id {}'.format(question['id'])))
+
+        if 'description' in question:
+            if self._search_dumb_quotes(question['description']):
+                errors.append(self._error_message(
+                    'Found dumb quotes in question description for question id {}'.format(question['id'])))
```
Comment (JamesGardiner) [inline app/validation/validator.py]:
```python
try:
    msg = f'Found dumb quotes in question description for question id {question_id}'
    description_error = self._error_message(msg) if self._search_dumb_quotes(question['description']) else None
except KeyError:
    pass

File: app/validation/validator.py (hunk 1 above)
Comment (berroar) [reply app/validation/validator.py]: Refactored with code from eq-translations

File: app/validation/validator.py [hunk 3]
```diff
@@ -1361,6 +1363,93 @@ def _validate_placeholder_transforms(self, transforms, block_id):
 
         return errors
 
+    def _validate_dumb_quotes_in_questions(self, question):
+        errors = []
+
+        if isinstance(question['title'], str):
+            question_title = question['title']
+        else:
+            question_title = question['title']['text']
```
[the remark this answers is not in the document]
Comment (berroar) [reply app/validation/validator.py]: Changed the solution, see Andrew's comment above.

File: app/validation/validate_schema_text.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
+import re
+from jsonpointer import resolve_pointer
+
+
+def get_pointer_keys():
```
Comment (iwootten) [inline app/validation/validate_schema_text.py]: I'm not sure duplicating some of the methods of the translations repo is the best approach here.

Given these are likely to change over time - it probably makes sense that eq-translations is installed as dependency here and then the relevant utility methods within it are used to locate pointers of interest. Much like our runner does to translate schemas at the moment.

Comment (berroar) [reply app/validation/validate_schema_text.py]: Have refactored and added eq-translations as a dependency

File: app/validation/validate_schema_text.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    list_pointers = schema_object.get_list_pointers()
+    list_contents = get_list_pointer_contents(json_schema, list_pointers)
+
+    for list_text in list_contents:
+        if search_dumb_quotes(list_text):
+            return error_message
+
+    for schema_text in schema_contents:
+        if search_dumb_quotes(schema_text):
+            return error_message
```
Comment (ajmaddaford) [inline app/validation/validate_schema_text.py]: - We should be generating an error for each string that contains a dumb quote. 
- We should create a single consolidated list of pointers, and then iterate over that.

Something like:

```python
pointers = []
errors = []

for key in get_pointer_keys():
    pointers.extend(find_pointers_to(json_schema, key))

schema_object = SurveySchema()
schema_object.load(schema_path)
pointers.extend(schema_object.get_list_pointers())

for pointer in pointers:
    schema_text = resolve_pointer(schema, pointer)
    if search_dumb_quotes(list_text):
        errors.append(f'Found dumb quote(s) in schema text at [{pointer}]')
```

File: app/validation/validate_schema_text.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+def search_dumb_quotes(text):
+    return re.search(r"'", text)
+
+
+def validate_json_schema_text(json_schema, schema_path):
+    error_message = 'Found dumb quotes in schema text'
+    schema_contents = get_pointer_contents(json_schema, get_pointer_keys())
+
+    schema_object = SurveySchema()
+    schema_object.load(schema_path)
```
Comment (ajmaddaford) [inline app/validation/validate_schema_text.py]: `schema_path` may not be needed, we may be able to do:

```python
schema_object = SurveySchema()
schema_object.schema = json_schema
```

File: app/validation/validate_schema_text.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+def get_list_pointer_contents(schema, schema_elements):
+    contents = []
+    for element in schema_elements:
+        pointer_contents = resolve_pointer(schema, element)
+        contents.append(pointer_contents)
+    return contents
+
+
+def search_dumb_quotes(text):
+    return re.search(r"'", text)
```
Comment (ajmaddaford) [inline app/validation/validate_schema_text.py]: If the other changes are made and this is only called from one place, I don't think this should be a separate method.

File: app/validation/validate_schema_text.py (hunk 5 above)
Comment (berroar) [reply app/validation/validate_schema_text.py]: Refactored this

File: app/validation/validate_schema_text.py (hunk 6 above)
Comment (berroar) [reply app/validation/validate_schema_text.py]: Fixed

File: app/validation/validate_schema_text.py (hunk 7 above)
Comment (berroar) [reply app/validation/validate_schema_text.py]: Removed the method

File: app/validation/validate_schema_text.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        'add_link_text',
+        'empty_list_text',
+        'text',
+        'title',
+    ]
+
+    return pointer_keys
+
+
+def get_list_pointer_contents(schema, schema_elements):
```
Comment (ajmaddaford) [inline app/validation/validate_schema_text.py]: This method is no longer used.

File: app/validation/validate_schema_text.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    for key in get_pointer_keys():
+        pointers.extend(find_pointers_to(json_schema, key))
+
+    schema_object = SurveySchema()
+    schema_object.schema = json_schema
+    pointers.extend(schema_object.get_list_pointers())
+
+    for pointer in pointers:
+        schema_text = resolve_pointer(json_schema, pointer)
+        if re.search(r"'", schema_text):
```
Comment (ajmaddaford) [inline app/validation/validate_schema_text.py]: The regex should be compiled before the for loop, so it's not done on every iteration:

```python
quote_regex = re.compile(r"'")
for pointer in pointers:
    schema_text = resolve_pointer(json_schema, pointer)
    if quote_regex.search(schema_text):
```

Also, the regex should also search for `"`.

Comment (berroar) [reply app/validation/validate_schema_text.py]: Fixed this

File: app/validation/validate_schema_text.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import re
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]: Theses imports need to be split to follow PEP8 standards

File: app/validation/validate_schema_text.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+from jsonpointer import resolve_pointer
+from eq_translations.utils import find_pointers_to
+from eq_translations.survey_schema import SurveySchema
+
+
+def get_pointer_keys():
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]:
```suggestion
```

File: app/validation/validate_schema_text.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        'playback',
+        'item_title',
+        'label',
+        'add_link_text',
+        'empty_list_text',
+        'text',
+        'title',
+    ]
+
+    return pointer_keys
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]:
```suggestion
```

File: app/validation/validate_schema_text.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def validate_json_schema_text(json_schema):
+    pointers = []
+    errors = []
+
+    for key in get_pointer_keys():
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]:
```suggestion
    pointers = [find_pointers_to(json_schema, key) for key in pointer_keys] + schema_object.get_list_pointers()
```

File: app/validation/validate_schema_text.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        'empty_list_text',
+        'text',
+        'title',
+    ]
+
+    return pointer_keys
+
+
+def validate_json_schema_text(json_schema):
+    pointers = []
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]:
```suggestion
```

File: app/validation/validate_schema_text.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    errors = []
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]:
```suggestion
```

File: app/validation/validate_schema_text.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    for key in get_pointer_keys():
+        pointers.extend(find_pointers_to(json_schema, key))
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]:
```suggestion
```

File: app/validation/validate_schema_text.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+    schema_object = SurveySchema()
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]: These can be moved to the start of the function

File: app/validation/validate_schema_text.py [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    schema_object.schema = json_schema
+    pointers.extend(schema_object.get_list_pointers())
```
Comment (JamesGardiner) [inline app/validation/validate_schema_text.py]:
```suggestion
```

File: tests/test_schema_text.py [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+from structlog import configure
+from structlog import getLogger
+from structlog.stdlib import LoggerFactory
+
+from app.validation.validate_schema_text import validate_json_schema_text
+
+logger = getLogger()
+
+configure(logger_factory=LoggerFactory())
```
Comment (JamesGardiner) [inline tests/test_schema_text.py]: Is the logger configuration needed? I don't see any log lines in the below code

File: tests/test_schema_text.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def validate_schema_text(json_to_validate):
+    errors = validate_json_schema_text(json_to_validate)
+    return errors
+
+
+def _open_and_load_schema_file(file):
+    json_file = open(os.path.join(os.path.dirname(os.path.realpath(__file__)), file), encoding='utf8')
```
Comment (JamesGardiner) [inline tests/test_schema_text.py]: use a context manager to open a file

File: tests/test_schema_text.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+    json_to_validate = load(json_file)
+
+    return json_to_validate
```
Comment (JamesGardiner) [inline tests/test_schema_text.py]: There isn't any point assigning the opened file - just return from within a context manager. The `__exit__` method will always be called and close the file handler correctly

File: tests/test_schema_text.py [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def test_param_valid_schemas(valid_schema_filename):
+    """
+    Uses py.test generated tests to validate all schemas contained in the 'valid' folder.
+    """
+    with open(valid_schema_filename, encoding='utf8') as json_file:
+        json_to_validate = load(json_file)
```
Comment (JamesGardiner) [inline tests/test_schema_text.py]: Any reason you aren't using the _open_and_load_schema_file function here

File: app/validation/validate_schema_text.py (hunk 12 above)
Comment (berroar) [reply app/validation/validate_schema_text.py]: Added this to the main function.

File: tests/test_schema_text.py (hunk 19 above)
Comment (berroar) [reply tests/test_schema_text.py]: I have removed these

File: tests/test_schema_text.py (hunk 20 above)
Comment (berroar) [reply tests/test_schema_text.py]: Amended the open and load function to use a context manager

File: tests/test_schema_text.py (hunk 22 above)
Comment (berroar) [reply tests/test_schema_text.py]: As above

File: app/validation/validate_schema_text.py (hunk 13 above)
Comment (berroar) [reply app/validation/validate_schema_text.py]: I don't think I am able to make this change as I need the return to be a flattened list of strings.

Comment (berroar) [reply app/validation/validate_schema_text.py]: No longer required after latest refactor

File: app/validation/validate_schema_text.py (hunk 18 above)
Comment (berroar) [reply app/validation/validate_schema_text.py]: Outdated

File: app/validation/validate_schema_text.py (hunk 14 above)
Comment (berroar) [reply app/validation/validate_schema_text.py]: Outdated

File: app/validation/validate_schema_text.py (hunk 16 above)
Comment (berroar) [reply app/validation/validate_schema_text.py]: Outdated

File: Pipfile [hunk 23]
```diff
@@ -8,10 +8,11 @@ name = "pypi"
 [packages]
 
 Flask = "*"
-jsonschema = "*"
+jsonschema = "<3"
 structlog = "*"
 python-dateutil = "*"
-
+jsonpointer = "*"
+eq-translations = {editable = true,git = "https://github.com/ONSDigital/eq-translations.git",ref = "master"}
```
Comment (JamesGardiner) [inline Pipfile]: Does eq-translations need to be installed with the editable flag? This has the effect of adding all the eq-translation dependencies to the Pipfile.lock

Comment (berroar) [reply Pipfile]: Removed this

File: app/validation/validator.py [hunk 24]
```diff
@@ -1361,6 +1365,22 @@ def _validate_placeholder_transforms(self, transforms, block_id):
...
+        schema_object.schema = json_schema
+        errors = []
+
+        # pylint: disable=invalid-string-quote
+        quote_regex = re.compile(r"['|\"]+(?![^{]*})+(?![^<]*>)")
+
+        for pointer in schema_object.pointers:
+            schema_text = resolve_pointer(json_schema, pointer)
+            if quote_regex.search(schema_text):
+                errors.append(self._error_message(f'Found dumb quotes(s) in schema text at [{pointer}]'))
```
Comment (iwootten) [inline app/validation/validator.py]: The square braces are a bit confusing here - they result in error messages that appear to be indicating lists.

Comment (berroar) [reply app/validation/validator.py]: I will remove these.

File: Dockerfile [hunk 25]
```diff
@@ -1,4 +1,8 @@
-FROM python:3.6-alpine
+FROM python:3.6-slim
```
Comment (JamesGardiner) [inline Dockerfile]: Does bumping to 3.8 break anything?

File: app/validation/validator.py [hunk 26]
```diff
@@ -5,6 +5,9 @@
 from datetime import datetime
 from json import load
 
+
+from jsonpointer import resolve_pointer
```
Comment (JamesGardiner) [inline app/validation/validator.py]: Blocks should be alphabetical

File: Dockerfile (hunk 25 above)
Comment (iwootten) [reply Dockerfile]: The Pipfile is based on 3.6, so presumably it wouldn't be installed?

Comment (JamesGardiner) [reply Dockerfile]: Sorry @berroar I should have marked this as a comment not a request for change.

Comment (JamesGardiner) [reply Dockerfile]: Sorry @berroar I should have marked this as a comment not a request for change.

File: app/validation/validator.py (hunk 26 above)
Comment (berroar) [reply app/validation/validator.py]: Fixed

**Comment by berroar** [comment]
Need to wait for https://github.com/ONSdigital/eq-survey-runner/pull/2486 before merging.

**Pull request merged**